lib/AST/NameLookup.cpp

```diff
@@ -1450,6 +1450,16 @@ static bool isAcceptableLookupResult(const DeclContext *dc,
     }
   }
 
+  // We don't support nested types in generics yet.
```

Comment:
I don't think "in generics" is accurate; you're filtering out generic type members from protocols and local context instead. However it would be better to allow such results, since we want to lift the restriction eventually (in particular I want to allow generic types in local context soon), and instead fix whatever other crash you were seeing.

Reply:
I don't know whether it's possible to fix without implementing that feature or it would actually cause even more fallout... 

Here is a reduced example which cases a crash:

```
class A : P {
  let d = A
}

protocol P {
  class A {}
}
```

In this example `lookupUnqualified` returns one inner result for `P.A` and one outer one for `A`. The problem is that https://github.com/apple/swift/blob/master/lib/Sema/TypeCheckNameLookup.cpp#L184L198 only handles associated type and protocol requirement cases. I was thinking about adding a `else` there to `{ addResult(found); return; }` but I'm not sure whether that's appropriate. 

This is yet another example of a situation which gets diagnosed by declaration checker and pushed on clients to handle (without even marking it as invalid).

Since I don't really know this area very well I'd be great to get an idea how do problems like this get fixed.

Reply:
Btw, I got "in generic" from `checkUnsupportedNestedType` check in `DeclChecker`...

Reply:
Ah, we want to get rid of that code eventually. Adding an 'else' there might work. If not, it looks like such results are simply dropped from the type checker lookup; which might be OK at that level too. I'm simply opposed to dropping them in the AST-level name lookup, since we shouldn't be enforcing semantic invariants there.

As for checkUnsupportedNestedType(), sounds like the comment there is out of date too :)

Reply:
Is adding that to LookupResultBuilder is fine, I'll gladly do that since it worked when I've testing it! :)